### PR TITLE
test: support regular expression matching on the response

### PR DESCRIPTION
**Pull request by ahrtr**
Currently we only support exact string match on the response (e.g. curl response, and process output). But in some cases, the response may not be a fixed string. One example is discussed in https://github.com/etcd-io/etcd/pull/16454, there may be random spaces in the grpc-gateway v2 response.

So we need to support match the response based on a regular expression.

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by ahrtr** [comment]
@fuweid @jmhbnz @serathius @spzala @wenjiaswe PTAL

**Review by jmhbnz — APPROVED** [review]
LGTM - Thanks @ahrtr.

Noting `expect.ExpectResponse` feels a bit verbose/duplicated naming  but I can't think of a better way of putting it in relation to the rest of the package.

File: pkg/expect/expect.go [hunk 1]
```diff
@@ -223,14 +229,23 @@ func (ep *ExpectProcess) ExpectFunc(ctx context.Context, f func(string) bool) (s
 }
 
 // ExpectWithContext returns the first line containing the given string.
-func (ep *ExpectProcess) ExpectWithContext(ctx context.Context, s string) (string, error) {
-	return ep.ExpectFunc(ctx, func(txt string) bool { return strings.Contains(txt, s) })
+func (ep *ExpectProcess) ExpectWithContext(ctx context.Context, s ExpectedResponse) (string, error) {
+	return ep.ExpectFunc(ctx, func(txt string) bool {
+		if s.IsRegularExpr {
```
Comment (fuweid) [inline pkg/expect/expect.go:234]: I think we can Compile out of closure and return error as soon as possible. And then we don't need to Compile it all the time.

File: tests/e2e/ctl_v3_kv_test.go [hunk 2]
```diff
@@ -354,25 +355,15 @@ func ctlV3Get(cx ctlCtx, args []string, kvs ...kv) error {
...
-	var lines []string
+	var lines []expect.ExpectedResponse
 	for _, elem := range kvs {
-		lines = append(lines, elem.key, elem.val)
+		lines = append(lines, expect.ExpectedResponse{Value: elem.key}, expect.ExpectedResponse{Value: elem.val})
 	}
 	return e2e.SpawnWithExpects(cmdArgs, cx.envMap, lines...)
 }
 
-// ctlV3GetWithErr runs "get" command expecting no output but error
```
Comment (fuweid) [inline tests/e2e/ctl_v3_kv_test.go:364]: Good catch!

File: pkg/expect/expect.go (hunk 1 above)
Comment (ahrtr) [reply pkg/expect/expect.go]: Makes sense & updated & thx

**Review by fuweid — APPROVED** [review]
LGTM

File: pkg/expect/expect.go [hunk 3]
```diff
@@ -223,14 +229,29 @@ func (ep *ExpectProcess) ExpectFunc(ctx context.Context, f func(string) bool) (s
...
+	return ep.ExpectFunc(ctx, func(txt string) bool {
+		if expr != nil {
+			return expr.MatchString(txt)
+		}
+		return strings.Contains(txt, s.Value)
+	})
 }
 
 // Expect returns the first line containing the given string.
 // Deprecated: please use ExpectWithContext instead.
```
Comment (wenjiaswe) [inline pkg/expect/expect.go:252]: This function is deprecated? Do we need to move it? Just want to mark, no need to remove in this PR,

File: pkg/expect/expect.go [hunk 4]
```diff
@@ -37,6 +38,11 @@ var (
 	ErrProcessRunning = fmt.Errorf("process is still running")
 )
 
+type ExpectedResponse struct {
+	Value         string
+	IsRegularExpr bool
```
Comment (wenjiaswe) [inline pkg/expect/expect.go:43]: Do you want to add some regex tests?

File: pkg/expect/expect.go (hunk 3 above)
Comment (ahrtr) [reply pkg/expect/expect.go:252]: We can cleanup the code in separate PR.

Comment (wenjiaswe) [reply pkg/expect/expect.go:252]: sgtm.

File: pkg/expect/expect.go (hunk 4 above)
Comment (ahrtr) [reply pkg/expect/expect.go:43]: added, please take a look at the second commit. thx

Comment (wenjiaswe) [reply pkg/expect/expect.go:43]: Thanks.

**Comment by wenjiaswe** [comment]
lgtm

**Comment by ahrtr** [comment]
thx @wenjiaswe for the quick review. Also thx @jmhbnz and @fuweid

**Pull request merged**